### PR TITLE
[CELEBORN-1252] Fix resource consumption of worker does not update when update interval is greater than heartbeat interval

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?

 Resource consumption of worker does not update when update interval of resource consumpution is greater than heartbeat interval.

截屏2024-01-24 14 49 50

This pull request introduces below changes:

1. Avoid master repeat add gauge for same user
2. For worker, user resource consumption can directly get from worker's snapshot, didn't need update interval 

### Why are the changes needed?

No.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

No.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
@SteNicholas Pls take a look since you add these metrics. cc @FMX

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 1]
```diff
@@ -534,49 +525,49 @@ private[celeborn] class Worker(
...
+          userIdentifier.toMap) { () =>
+          if (workerInfo.userResourceConsumption.containsKey(userIdentifier)) {
+            workerInfo.userResourceConsumption.get(userIdentifier).hdfsBytesWritten
+          } else {
+            0L
+          }
         }
       }
     }
     workerInfo.updateThenGetUserResourceConsumption(resourceConsumptionSnapshot.asJava)
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:568]: when workerInfo updateThenGetUserResourceConsumption,  it will first clear userResourceConsumption, then it may cause user resource consumption metrics to 0.

Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:568]: > when workerInfo updateThenGetUserResourceConsumption, it will first clear userResourceConsumption, then it may cause user resource consumption metrics to 0.

Nice catch, it's a problem.

Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:568]: @RexXiong How about current?

Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:568]: How about current?

**Review by RexXiong — APPROVED** [review]
LGTM, thanks!

File: common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala [hunk 2]
```diff
@@ -188,11 +188,14 @@ class WorkerInfo(
     JavaUtils.newConcurrentHashMap[String, DiskInfo](diskInfos)
   }
 
-  def updateThenGetUserResourceConsumption(consumption: util.Map[
+  def updateThenGetUserResourceConsumption(userConsumption: util.Map[
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:191]:
```suggestion
  def updateThenGetUserResourceConsumption(resourceConsumptions: util.Map[
```

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 3]
```diff
@@ -877,7 +877,6 @@ private[celeborn] class Master(
   private def handleCheckQuota(
       userIdentifier: UserIdentifier,
       context: RpcCallContext): Unit = {
-
```
Comment (SteNicholas) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:880]: Revert this change.

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 4]
```diff
@@ -898,7 +897,6 @@ private[celeborn] class Master(
       userIdentifier.toMap) { () =>
       computeUserResourceConsumption(userIdentifier).hdfsBytesWritten
     }
-
```
Comment (SteNicholas) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:901]: Ditto.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

**Comment by SteNicholas** [comment]
@AngersZhuuuu, could you update `migration.md` to describe the removal of `celeborn.worker.userResourceConsumption.update.interval`?

**Review by SteNicholas — APPROVED** [review]

File: common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala [hunk 5]
```diff
@@ -188,11 +188,14 @@ class WorkerInfo(
     JavaUtils.newConcurrentHashMap[String, DiskInfo](diskInfos)
   }
 
-  def updateThenGetUserResourceConsumption(consumption: util.Map[
+  def updateThenGetUserResourceConsumption(resourceConsumptions: util.Map[
     UserIdentifier,
     ResourceConsumption]): util.Map[UserIdentifier, ResourceConsumption] = {
-    userResourceConsumption.clear()
-    userResourceConsumption.putAll(consumption)
+    userResourceConsumption.keys().asScala.filterNot(resourceConsumptions.containsKey).foreach {
```
Comment (FMX) [inline common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:194]: I think this change is incorrect.  According to this logic, If an identifier is gone, there will always be zero resource consumption attached to it.
I think make the "clear" and "putall" in a atomic block should be ok instead of this change.

Comment (FMX) [reply common/src/main/scala/org/apache/celeborn/common/meta/WorkerInfo.scala:194]: I think just removing the inexistent idenfiers from user resource consumption will be better.

**Review by FMX — APPROVED** [review]
LGTM. I think there will be a patch to clean outdated metrics in the future.
